builder/openstack/step_run_source_server.go

```diff
@@ -150,16 +152,34 @@ func (s *StepRunSourceServer) Cleanup(state multistep.StateBag) {
 		return
 	}
 
+	maxNumErrors := 10
+	numErrors := 0
+
 	ui.Say(fmt.Sprintf("Terminating the source server: %s ...", s.server.ID))
-	if config.ForceDelete {
-		if err := servers.ForceDelete(computeClient, s.server.ID).ExtractErr(); err != nil {
-			ui.Error(fmt.Sprintf("Error terminating server, may still be around: %s", err))
-			return
+	for {
+		if config.ForceDelete {
+			err = servers.ForceDelete(computeClient, s.server.ID).ExtractErr()
+		} else {
+			err = servers.Delete(computeClient, s.server.ID).ExtractErr()
 		}
-	} else {
-		if err := servers.Delete(computeClient, s.server.ID).ExtractErr(); err != nil {
-			ui.Error(fmt.Sprintf("Error terminating server, may still be around: %s", err))
-			return
+
+		if err != nil {
+			_, OK := err.(gophercloud.ErrDefault500)
+			if OK {
+				numErrors++
+				if numErrors >= maxNumErrors {
+					ui.Error(fmt.Sprintf("Error terminating server, maximum number (%d) reached: %s", numErrors, err))
+					return
+				}
+				log.Printf("Error terminating server on (%d) time(s): %s, retrying ...", numErrors, err)
+				time.Sleep(2 * time.Second)
+				continue
+			} else {
+				ui.Error(fmt.Sprintf("Error terminating server, may still be around: %s", err))
+				return
+			}
+		} else {
+			break
```

Comment:
Hey @zhenggu I had a little trouble immediately understanding what is going on. I suggest a slight refactor to help with readability. Could you validate that the suggested change works as intended please. 
```go
	for {
		if config.ForceDelete {
			err = servers.ForceDelete(computeClient, s.server.ID).ExtractErr()
		} else {
			err = servers.Delete(computeClient, s.server.ID).ExtractErr()
		}

		if err == nil {
			break
		}

		if _, ok := err.(gophercloud.ErrDefault500); !ok {
			ui.Error(fmt.Sprintf("Error terminating server, may still be around: %s", err))
			return
		}

		if numErrors < maxNumErrors {
			numErrors++
			log.Printf("Error terminating server on (%d) time(s): %s, retrying ...", numErrors, err)
			time.Sleep(2 * time.Second)
			continue
		}
		ui.Error(fmt.Sprintf("Error terminating server, maximum number (%d) reached: %s", numErrors, err))
		return
	}
```